### ISSUE TITLE
Consider changing `Dict[str,any]` of generic fields to `any` in `Calibration` model

### ISSUE THREAD

**Issue by bruno-f-cruz** [AllenNeuralDynamics/aind-data-schema#692]
**Is your feature request related to a problem? Please describe.**
Some fields across the schema are typed as `Dict[str, any]` (e.g. `Calibration` `input`) to allow users to flexibility pack custom data structures into the field. Unfortunately, while the current type has indeed a lot of expressivity, it leads to clunky notation if a user wants to plug a custom pydantic model (e.g. inherited from `AindModel`) into the field directly. The current way would be `Dict[str, MyModel]`. Where the key argument would be a bit unnecessary and even confusing, since the end-user would need to know what keys are available by querying the dictionary first.

**Describe the solution you'd like**
I would suggest that changing the type to `any` might make more sense since it would allow for the current typing (i.e. it would be completely backward compatible) while at the same time allowing other classes to be used. This is critical if the aind-data-schema is to be adopted by users that might want to build their own schemas while inheriting from the available core classes.

**Describe alternatives you've considered**
Leave it as is and live with the current alternative.

**Additional context**
See discussion https://github.com/AllenNeuralDynamics/aind-data-schema/discussions/636#discussioncomment-7805135 for a lengthier example

**Comment by saskiad** [comment]
@jtyoung84 @dyf Do you have thoughts on this?

**Comment by jtyoung84** [comment]
I guess it depends on the expected use case. We can restrict it to `Dict[str, "primitive types like int, str, float, array, etc."]` and require users to work within those restrictions. We can also use Generics, but that gets complicated. If we really want this to be a field without any restrictions on the type, then we can change it to `Any`.

**Comment by saskiad** [comment]
I have a hard time envisioning a situation where we don't use a string as a the key for the Dict. Is that really something you need Bruno? That makes me nervous, honestly.

**Comment by bruno-f-cruz** [comment]
Say I want to pass the following model:

```python

class MyCalibrationStructure(AindModel):
    slope: float
    offset: float

```

If we don't change anything we have two options:

1. Put something random in the key of the dictionary like "Value". This will give you the ability to directly inherit the `Calibration` class and add a used-specified model to constrain (and type) the `input` property. The user still has the ability to use their own schema, since the type is still `Any` so its ok. Bad news, a downstream user reading the schema must know what that key is, best case scenario by grabbing all keys of the dictionary.

2. Let users create their own schema but force them to serialize all classes into `Dict[str, any]` before populating `Calibration`, by calling Model.model_dump(). This allows you to keep using your current type hinting BUT adds the step on the user side.

The difference is subtle but important Imo. One affords/incentivizes users to build their own schemas within the spirit of what I think aind-data-schema is trying to achieve. The other constrains the input because it assumes users might not want to build their own schemas and are ok with coercing everything into a generic dictionary. 

In code you would get:

```python

from aind_data_schema.models.devices import Calibration
from aind_data_schema.base import AindModel
from typing import Any, Dict, Literal
import datetime


## With Any as the input

calibration_date = datetime.datetime.now()

class MyCalibrationStructure(AindModel):
    slope: float
    offset: float


class CalibrationChild(Calibration):
    device_name: Literal["Valve"] = "Valve"
    input: list[float]
    output: MyCalibrationStructure


print(
    CalibrationChild(
        calibration_date=calibration_date,
        input=[1, 2, 3],
        output=MyCalibrationStructure(slope=1, offset=2),
        description="",
    )
)

# Keeping the current Dict[str, Any] as the input and the user uses their own model

class MyCalibrationStructure(AindModel):
    slope: float
    offset: float


class CalibrationChild(Calibration):
    device_name: Literal["Valve"] = "Valve"
    input: Dict[str, list[float]]
    output: Dict[str, MyCalibrationStructure]


print(
    CalibrationChild(
        calibration_date=calibration_date,
        input={"Foo": [1, 2, 3]},
        output={"Bar": MyCalibrationStructure(slope=1, offset=2)},
        description="",
    )
)

# Keeping the current Dict[str, Any] as the input and the user serializes their model

class MyCalibrationStructure(AindModel):
    slope: float
    offset: float


class CalibrationChild(Calibration):
    device_name: Literal["Valve"] = "Valve"


print(
    CalibrationChild(
        calibration_date=calibration_date,
        input={"Foo": [1, 2, 3]},
        output=MyCalibrationStructure(slope=1, offset=2).model_dump(),
        description="",
    )
)
```

**Comment by bruno-f-cruz** [comment]
(I seem to have deleted the previous post, sorry)

Here's the difference in code.

tldr: If you keep it as is, you either have a clunky notation or you push users to validate everything on their side and simply push a serialized dictionary to your field.

If we make it Any, we can still have any of the above, but also pass a pedantic model directly. I believe this is very much within the spirit of letting people inherit from the aind-data-schema and build their own models, which I would encourage if we want a wide adoption and more contributors.

```python from aind_data_schema.models.devices import Calibration
from aind_data_schema.base import AindModel
from typing import Any, Dict, Literal
import datetime


## With Any as the input

calibration_date = datetime.datetime.now()

class MyCalibrationStructure(AindModel):
    slope: float
    offset: float


class CalibrationChild(Calibration):
    device_name: Literal["Valve"] = "Valve"
    input: list[float]
    output: MyCalibrationStructure


print(
    CalibrationChild(
        calibration_date=calibration_date,
        input=[1, 2, 3],
        output=MyCalibrationStructure(slope=1, offset=2),
        description="",
    )
)

# Keeping the current Dict[str, Any] as the input and the user uses their own model

class MyCalibrationStructure(AindModel):
    slope: float
    offset: float


class CalibrationChild(Calibration):
    device_name: Literal["Valve"] = "Valve"
    input: Dict[str, list[float]]
    output: Dict[str, MyCalibrationStructure]


print(
    CalibrationChild(
        calibration_date=calibration_date,
        input={"Foo": [1, 2, 3]},
        output={"Bar": MyCalibrationStructure(slope=1, offset=2)},
        description="",
    )
)

# Keeping the current Dict[str, Any] as the input and the user serializes their model

class MyCalibrationStructure(AindModel):
    slope: float
    offset: float


class CalibrationChild(Calibration):
    device_name: Literal["Valve"] = "Valve"


print(
    CalibrationChild(
        calibration_date=calibration_date,
        input={"Foo": [1, 2, 3]},
        output=MyCalibrationStructure(slope=1, offset=2).model_dump(),
        description="",
    )
)
```

**Comment by bruno-f-cruz** [comment]
Heya. Can we review this issue? Is this still on the table or will it become a wontfix? It would be nice to know asap to figure out how to maintain things on my side. Thanks!

**Comment by jtyoung84** [comment]
Here are the options as I see it:
 
1) Use `Dict[str, Any]`, but then I kind of understand where Bruno might be coming from
2) Use `Any`, which will be really difficult to de-serialize things from JSON
3) Use `Dict[str, primitive types]`. This is good if the field isn't expected to be too complicated
4) Use a tagged `Union` like we do with other fields that have multiple Model types

I lean towards 3 or 4, but I don't feel too strongly if we decide to stick with 1 or 2.

**Comment by bruno-f-cruz** [comment]
3 is moving in the opposite direction of the proposal I feel. I like 4 but
it means that the models that I want to use must be imported in
aind-data-schemas.

How about a mix between 4 and 2 where, instead of Any, we use AindModel as
input?

**Comment by bruno-f-cruz** [comment]
Can you give an example why you think 3. Would it make difficult to deserialize the json? I am struggling a bit with that point.

**Comment by jtyoung84** [comment]
> Can you give an example why you think 2. Would make it difficult to deserialize the json? I am struggling a bit with that point. I don't see how, worst case scenario, it would be any different than it is now. But maybe I am missing something?

```
from pydantic import BaseModel
from typing import Any

class ModelA(BaseModel):
   id: int = 1
   name: str = "abc"

class ModelB(BaseModel):
   input: Any

m = ModelB(input=ModelA())
m_json = m.model_dump_json()
m2 = ModelB.model_validate_json(m_json)
print(m2 == m)
# False
```

**Comment by bruno-f-cruz** [comment]
I dont think that will ever be possible (by design) since the base class doesn't know about the sub-class. I think the test should rather be:

```python
from pydantic import BaseModel, Field
from typing import Any

class ModelA(BaseModel):
   id: int = 1
   name: str = "abc"

class ModelB(BaseModel):
   input: Any

class Container(BaseModel):
   model: Any

class ModifiedContainer(BaseModel):
   model: ModelA = Field(ModelA())


m = Container(model=ModifiedContainer())
m_json = m.model_dump_json()
print(m_json)
# test if coercing through the base class destroys the base class round-trip
m2 = ModifiedContainer.model_validate_json(m_json)
print(m2 == ModifiedContainer())

```

It should be noted that m still has a dict-like representation of the object that can be used (similarly to what already happens currently, as key:value pairs).
Indeed, for types known natively, your requirement actually works:

```python
dict_test = {"param1": 1, "param2": "abc"}

m = Container(model=dict_test)
m_json = m.model_dump_json()
m2 = Container.model_validate_json(m_json)
print(m2 == m)

```


Does this make sense?

**Comment by jtyoung84** [comment]
For the use case you're describing, I've seen most people use Generics:
```
from pydantic import BaseModel
from typing import Generic, TypeVar

T = TypeVar('T')

class ModelA(BaseModel):
   id: int = 1
   name: str = "abc"

class ModelB(BaseModel, Generic[T]):
   input: T

m = ModelB[ModelA](input=ModelA())
m_json = m.model_dump_json()
m2 = ModelB[ModelA].model_validate_json(m_json)
print(m2 == m)
# True
```

**Comment by bruno-f-cruz** [comment]
I like the idea but it still has the same problem as i described before. The aind-data-schema will have to know about the custom  ModelA type to perform the correct "casting". That being said, I can totally work with this I think. Just not sure what it buys you versus the Any in practice. 
Happy to go with this, i just tested it and it seems to be easy to subclass:

```python
from pydantic import BaseModel, Field
from typing import Any
from typing import Generic, TypeVar
T = TypeVar('T')


class ModelA(BaseModel):
   id: int = 1
   name: str = "abc"

class Container(BaseModel, Generic[T]):
   model: T

class ModifiedContainer(Container):
   model: ModelA = Field(ModelA())

```

Is this what you had in mind?

**Comment by jtyoung84** [comment]
Another alternative would be to declare the field as an unconstrained JSON type (https://docs.pydantic.dev/latest/api/types/#pydantic.types.Json). The round-trip should evaluate correctly. Additional work may be required to parse the json string into a specific pydantic model.

**Comment by bruno-f-cruz** [comment]
I think these are all valid solutions, but I think it boils down to these two questions:

 - Are the aind-data-schema intended to be extended by users, or to be used as is? 
 - If to be extended, what should be the supported pattern (this issue)
 
Depending on the answers to these two points, I think the solution will be rather easy to make. 

Stepping back, this is a critical decision to be made. As a user, I would like to know whether to invest time adhering to the current specs, or simply move forward with an isolated set of schemas of my own. Given limited time resources, I feel that asking users to code things twice (their own schemas and the conversion layer) seems a bit counter-productive, so my preference would be to kill two birds with one stone. 

If you want to discuss this in person lmk @saskiad @dyf @jtyoung84

**Comment by bruno-f-cruz** [comment]
Would this satisfy all the requirements?


```python

from aind_data_schema.base import AindCoreModel, AindModel
from pydantic import BaseModel, RootModel, Field, ValidationError
from typing import List, Optional, Union, Dict, Any, Literal, TypeVar, Generic, Annotated
import json

AindGenericType = TypeVar("AindGenericType", bound="AindGeneric") 

class AindGeneric(BaseModel, extra="allow"):
    pass
    # We can use this class to add some metadata for classes we think might be subclassed in the future. It could contain a sort of "describedBy" that could allow readers to fetch the schema of this subclass if needed. Otherwise just leave this as an abstract class.

class AindDataSchemaModel(AindCoreModel, Generic[AindGenericType]): # This would be the class as defined in the aind-data-schema
    describedBy: Literal["NA"] = "NA"
    schema_version: Literal["0.1.0"] = "0.1.0"
    input: AindGenericType
    name: str = "name"

# This would be what an user might want to inject in the Generic field
class MyCustomInput(BaseModel):
    param1: int
    param2: str


class MyCustomModel(AindDataSchemaModel): #  This would be the pattern of inheritance.
    input: MyCustomInput = Field(MyCustomInput, validate_default=True)

# making sure the schema looks good...
with open("data.json", "w") as f:
    model = (AindDataSchemaModel.model_json_schema())
    f.write(json.dumps(model, indent=2))

# Start with a dictionary and round-trip

data = AindDataSchemaModel(input={"foo": "bar", "baz": 1})
data_json = data.model_dump_json()
data_prime = AindDataSchemaModel.model_validate_json(data_json)
print(data)
print(data_prime)
print(data == data_prime)

# Fail with a non-allowed type
try:
    data = AindDataSchemaModel(input=1)
    data_json = data.model_dump_json()
    data_prime = AindDataSchemaModel.model_validate_json(data_json)
    print(data)
    print(data_prime)
    print(data == data_prime)
except ValidationError as e:
    print(e)


# Ensure generic deserialization behavior from the base class while retaining equality in the subclass
data = MyCustomModel(input=MyCustomInput(param1=2, param2="bar"))
data_json = data.model_dump_json()
print(data_json)
data_prime = AindDataSchemaModel.model_validate_json(data_json)
data_prime_json = data_prime.model_dump_json()
print(data_prime_json)
recovered_model = MyCustomModel.model_validate_json(data_prime_json)
print(data == data_prime)
print(data == recovered_model)
```

**Comment by jtyoung84** [comment]
This looks like it might work. The Calibration would look something like:
```python
class Calibration(AindModel, Generic[AindGenericType]):
    """Generic calibration class"""
    calibration_date: datetime = Field(..., title="Date and time of calibration")
    device_name: str = Field(..., title="Device name", description="Must match a device name in rig/instrument")
    description: str = Field(..., title="Description", description="Brief description of what is being calibrated")
    input: AindGenericType = Field(AindGeneric(), description="Calibration input", title="inputs")
    output: Dict[str, Any] = Field(dict(), description="Calibration output", title="outputs")
    notes: Optional[str] = Field(None, title="Notes")
```

**Comment by bruno-f-cruz** [comment]
This looks like it might work. The Calibration would look something like:
```python
class Calibration(AindModel, Generic[AindGenericType]):
    """Generic calibration class"""
    calibration_date: datetime = Field(..., title="Date and time of calibration")
    device_name: str = Field(..., title="Device name", description="Must match a device name in rig/instrument")
    description: str = Field(..., title="Description", description="Brief description of what is being calibrated")
    input: AindGenericType = Field(AindGeneric(), description="Calibration input", title="inputs")
    output: AindGenericType = Field(AindGeneric(), description="Calibration output", title="outputs")
    notes: Optional[str] = Field(None, title="Notes")
```

Yep. I think we should replace all of the previous signature entries (i.e. including `output`) for consistency. Another thing that your example made me realize is that we should have AindGeneric() as a default indeed. 

I would leave it up to the specific aind-data-schema to decide whether the field should be required and/or nullable. In certain cases I could see Optional[AindGenericType] making sense.

I am happy to go with this solution!

**Comment by dyf** [comment]
this is done
